File: src/my_odometry/scripts/live_plot.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include "matplotlibcpp.h"
#include <mutex>
#include <thread>
#include <atomic>
#include <vector>
#include <chrono>

namespace plt = matplotlibcpp;

class LivePlot : public rclcpp::Node {
public:
    LivePlot() : Node("live_plot"), running_(true) {
        // ROS2 subscriptions
        imu_sub_ = create_subscription<sensor_msgs::msg::Imu>(
            "/imu_points", 10, std::bind(&LivePlot::imu_cb, this, std::placeholders::_1));
        odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
            "/odom_points", 10, std::bind(&LivePlot::odom_cb, this, std::placeholders::_1));

        // Matplotlib interactive mode
        plt::backend("Qt5Agg");
        plt::ion();
        plt::figure_size(800, 600);
        plt::xlabel("X");
        plt::ylabel("Y");
        plt::title("Live Odometry + IMU Plot");

        // Start plotting thread
        plot_thread_ = std::thread([this]() { this->plot_loop(); });
    }

    ~LivePlot() {
        running_ = false;
        if (plot_thread_.joinable()) {
            plot_thread_.join();
        }
    }

private:
    void imu_cb(const sensor_msgs::msg::Imu::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(mutex_);
        imu_x_.push_back(msg->linear_acceleration.x);
        imu_y_.push_back(msg->linear_acceleration.y);

        // Keep vectors from growing indefinitely (optional)
        if (imu_x_.size() > max_points_) {
            imu_x_.erase(imu_x_.begin());
            imu_y_.erase(imu_y_.begin());
        }
    }

    void odom_cb(const nav_msgs::msg::Odometry::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(mutex_);
        odom_x_.push_back(msg->pose.pose.position.x);
        odom_y_.push_back(msg->pose.pose.position.y);

        if (odom_x_.size() > max_points_) {
            odom_x_.erase(odom_x_.begin());
            odom_y_.erase(odom_y_.begin());
        }
    }

    void plot_loop() {
        while (rclcpp::ok() && running_) {
            {
                std::lock_guard<std::mutex> lock(mutex_);
                plt::clf();
                if (!imu_x_.empty()) plt::plot(imu_x_, imu_y_, "r-");
                if (!odom_x_.empty()) plt::plot(odom_x_, odom_y_, "b-");
                plt::legend();
                plt::draw();
                plt::pause(0.05);
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }
    }

    // ROS2 subscriptions
    rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr imu_sub_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;

    // Data storage
    std::vector<double> imu_x_, imu_y_;
    std::vector<double> odom_x_, odom_y_;
    std::mutex mutex_;
    const size_t max_points_ = 1000;  // limit stored points

    // Plotting thread
    std::thread plot_thread_;
    std::atomic<bool> running_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LivePlot>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_odometry/scripts/live_plot.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
import matplotlib.pyplot as plt
import threading

class LivePlot(Node):
    def __init__(self):
        super().__init__('live_plot')
        self.enc_x, self.enc_y = [], []
        self.imu_x, self.imu_y = [], []

        self.sub_enc = self.create_subscription(
            Odometry, '/odom_est', self.enc_cb, 10)
        self.sub_imu = self.create_subscription(
            Odometry, '/odom_imu_est', self.imu_cb, 10)

        # start a background thread for matplotlib
        self.plot_thread = threading.Thread(target=self.plot_loop, daemon=True)
        self.plot_thread.start()

    def enc_cb(self, msg):
        self.enc_x.append(msg.pose.pose.position.x)
        self.enc_y.append(msg.pose.pose.position.y)

    def imu_cb(self, msg):
        self.imu_x.append(msg.pose.pose.position.x)
        self.imu_y.append(msg.pose.pose.position.y)

    def plot_loop(self):
        plt.ion()
        fig, ax = plt.subplots()
        while rclpy.ok():
            ax.clear()
            ax.plot(self.enc_x, self.enc_y, 'g-', label='Encoders')
            ax.plot(self.imu_x, self.imu_y, 'b-', label='IMU')
            ax.set_xlabel('X [m]')
            ax.set_ylabel('Y [m]')
            ax.legend()
            plt.pause(0.05)


def main(args=None):
    rclpy.init(args=args)
    node = LivePlot()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

